### PR TITLE
Sequential lineage tracing data simulator

### PR DESCRIPTION
**Pull request by colganwi**
This PR adds a `SequentialLineageTracingDataSimulator` class for modeling sequential lineage tracing systems such as the DNA Typewriter described by [Choi et al, Nature 2022](https://www.nature.com/articles/s41586-022-04922-8)

### CODE REVIEW

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                                / self.continuation_rate)
+                            if time < life_time:
+                                self.edit_site(
+                                    character_array, i, self.state_priors
+                                )
+                        i += 1
+                # uninitiated
+                elif character_array[c] == 0:
+                    i = c 
+                    time = -np.log(np.random.uniform()) / self.initiation_rate
```
Comment (mattjones315) [inline cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:187]: Sorry, perhaps I don't fully understand this math, but why not just use `np.random.exponential(self.initiation_rate)` to draw a time to the next event?

**Review by mattjones315 — CHANGES_REQUESTED** [review]
Looks great! Just a few comments here to address before merging in

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+        if number_of_cassettes <= 0 or not isinstance(number_of_cassettes, int):
+            raise DataSimulatorError("Specify a positive number of cassettes.")
+        if size_of_cassette <= 0 or not isinstance(size_of_cassette, int):
+            raise DataSimulatorError(
+                "Specify a positive number of cut-sites per cassette."
+            )
+        if not isinstance(initiation_rate, float):
+            raise DataSimulatorError("Initiation rate must be a float.")
+        if initiation_rate <= 0:
```
[the remark this answers is not in the document]
Comment (colganwi) [reply cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:104]: I don't understand the question. This function raises an error if `initial_rate` or `continuation_rate` is 0

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py (hunk 1 above)
Comment (colganwi) [reply cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:187]: Good point

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattjones315) [reply cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:104]: Sorry, you're right - I misread the initial logic flow...

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
+"""
+A Sequential lineage tracing data simulator. This is a subclass of the
+LineageTracingDataSimulator that simulates the data produced from sequential
+Cas9-based technologies (e.g, Typewriter described in Choi et al, Nature 2022). 
+This simulator implements the method `overlay_data` which takes
+in a CassiopeiaTree with edge lengths and overlays states.
+"""
+import copy
```
Comment (mattjones315) [inline cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:8]: a couple of imports don't seem to be used here -- `copy`, `Callable`, `Tuple`, and `Union`. Let's remove them if they're not used.

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                # missing
+                if character_array[c] == -1:
+                    continue
+                # initiated
+                elif character_array[c] != 0:
+                    time = 0
+                    i = c + 1
+                    while time < life_time and i < c + self.size_of_cassette:
+                        if character_array[i] == 0:
+                            time = time + (-np.log(np.random.uniform())
```
Comment (mattjones315) [inline cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:177]: sorry again if I misunderstand the math, but couldn't this also just be drawn from an exponential as below, `time += np.random.exponential(1/self.continuation_rate)`?  Or is this a different waiting time distribution?

**Review by mattjones315 — CHANGES_REQUESTED** [review]
Thanks for responding the previous requests, I have a few more. Would you mind addressing these comments?

File: cassiopeia/simulator/SequentialLineageTracingDataSimulator.py (hunk 4 above)
Comment (colganwi) [reply cassiopeia/simulator/SequentialLineageTracingDataSimulator.py:177]: Thanks! Missed that one

**Review by mattjones315 — APPROVED** [review]
Looks great! Thanks for addressing these comments. Merging it in now.

**Pull request merged**